### PR TITLE
feat(ui): implemented full student dashboard

### PR DESCRIPTION
**Pull request by mostafakamar2308**
For the past and upcoming lessons are these the correct queries?
`UpcomingLessons`:
image

`pastLessons`:
image

The result:

image

### CODE REVIEW

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-mug30un8l-litespace-core.vercel.app) 
🚀 [Luna](https://luna-9hbx8nzj7-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-4vp7a3k7g-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-dtv3fhute-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-i0ozsrq2l-litespace-core.vercel.app) 
🚀 [Luna](https://luna-4h3vkhkc9-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-dyccdn6za-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-5ckyludog-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-kf8mbwo62-litespace-core.vercel.app) 
🚀 [Luna](https://luna-imcawneaj-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-8gj7s77e7-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-aulit5xj3-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-mlxbs1b7h-litespace-core.vercel.app) 
🚀 [Luna](https://luna-54vurqih7-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-3qp4bh1lr-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-3o9yok1jw-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-72ft978ms-litespace-core.vercel.app) 
🚀 [Luna](https://luna-pb33aaf5k-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-ohkdcsfjh-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-78cv10qie-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-qaq6uuafx-litespace-core.vercel.app) 
🚀 [Luna](https://luna-rbcxr4ujp-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-qxga6xnaf-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-dnyq2woic-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-fdosp340b-litespace-core.vercel.app) 
🚀 [Luna](https://luna-blorj5b6d-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-9rga42zbr-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-1wgby3gfs-litespace-core.vercel.app)

**Comment by neuodev** [comment]
image

- The alignment of the elements is not right.  They are not at the same line. 
- The card width doesn't match the design.

File: packages/luna/src/components/StudentOverview/StudentOverview.tsx [hunk 1]
```diff
@@ -48,11 +48,13 @@ export const StudentOverview: React.FC<Props> = ({
...
+
     if (totalLearningTime < 60 * 1000)
-      return formatMinutes(totalLearningTime, {
-        units: ["h"],
-        maxDecimalPoints: 0,
-      });
+      return duration.hours({ language: "ar-short" });
 
     const hours = Math.floor(totalLearningTime / 60);
     return intl("student-dashboard.overview.total-learning-time.unit", {
```
Comment (neuodev) [inline packages/luna/src/components/StudentOverview/StudentOverview.tsx:60]: Based on your changes, we should update this logic as well. to use `duration.format`

Comment (neuodev) [reply packages/luna/src/components/StudentOverview/StudentOverview.tsx:60]: The method named `duration.hours` is misleading. We should use `duration.format` and modifiy to be customizable to accept options.

**Review by neuodev — CHANGES_REQUESTED** [review]

File: apps/nova/src/components/Common/Loader.tsx [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        <div className="w-[60px] h-[60px] bg-white rounded-full" />
+        <div className="w-[13px] h-[10px] bg-[rgba(29,124,78,1)] rounded-full absolute bottom-0 left-1/2 -translate-x-1/2" />
+      </div>
+
+      <Typography element="h4" weight="bold" className="text-natural-950">
+        {text}
+      </Typography>
+    </div>
+  );
+};
```
[the remark this answers is not in the document]
Comment (mostafakamar2308) [reply apps/nova/src/components/Common/Loader.tsx:19]: yes, you are right

File: apps/nova/src/components/dashboard/PastLessonsWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        duration: lesson.lesson.duration,
+        tutor: {
+          id: teacher?.userId || 0,
+          name: teacher?.name || null,
+          imageUrl: teacher?.image || null,
+        },
+      };
+    }) || []
+  );
+}
```
[the remark this answers is not in the document]
Comment (mostafakamar2308) [reply apps/nova/src/components/dashboard/PastLessonsWrapper.tsx:34]: this is cleaner 🚀

File: packages/luna/src/components/StudentOverview/StudentOverview.tsx (hunk 1 above)
Comment (mostafakamar2308) [reply packages/luna/src/components/StudentOverview/StudentOverview.tsx:60]: yes I changed the formate function.

I didn't change the logic of line `59` as humanize doesn't formate numbers so i left it as it's

File: packages/luna/src/components/Lessons/PastLessonsTable.tsx [hunk 4]
```diff
@@ -85,7 +85,7 @@ export const PastLessonsTable: React.FC<Props> = ({
               </div>
               <div className="tw-mb-4">
                 <div className="tw-flex tw-flex-row tw-gap-2 tw-items-center">
-                  <div className="tw-w-10 tw-h-10 tw-rounded-full tw-overflow-hidden">
+                  <div className="tw-w-10 tw-h-10 tw-rounded-lg tw-overflow-hidden">
```
[the remark this answers is not in the document]
Comment (mostafakamar2308) [reply packages/luna/src/components/Lessons/PastLessonsTable.tsx:88]: but It won't benefit from it, or do you want me to move the logic in the `pastLessonTable`?

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-oo1jfehl2-litespace-core.vercel.app) 
🚀 [Luna](https://luna-f9byh8fgk-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-85j4hexga-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-fbpr1lwk6-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-fnmu64dpy-litespace-core.vercel.app) 
🚀 [Luna](https://luna-2vr3hhmw7-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-732lcbw6c-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-6sw6msqxl-litespace-core.vercel.app)

File: apps/nova/src/components/dashboard/ChatSummaryWrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      id: room.roomId,
+      url: Route.Chat.concat("?room=", room.roomId.toString()),
+      name: orUndefined(room.otherMember?.name),
+      image: orUndefined(room.otherMember?.image),
+      message: room.latestMessage?.text || "TODO",
+      sentAt: room.latestMessage?.updatedAt || dayjs().toString(), // TODO
+      read: room.unreadMessagesCount === 0,
+    })) || []
+  );
+}
```
Comment (neuodev) [inline apps/nova/src/components/dashboard/ChatSummaryWrapper.tsx:24]: The code can be a little bit cleaner by doing the following: 

```suggestion
function asChatRooms(
  list: IRoom.FindUserRoomsApiRecord[] | null
): ChatSummaryProps["rooms"] {
  if (!list) return [] // adding safeguard will make the code more cleaner. 
  return (
    list.map((room) => ({
      id: room.roomId,
      url: Route.Chat.concat("?room=", room.roomId.toString()),
      name: orUndefined(room.otherMember?.name),
      image: orUndefined(room.otherMember?.image),
      message: room.latestMessage?.text || "TODO",
      sentAt: room.latestMessage?.updatedAt || dayjs().toString(), // TODO
      read: room.unreadMessagesCount === 0,
    }));
  );
}
```

File: apps/nova/src/components/dashboard/PastLessonsWrapper.tsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  const intl = useFormatMessage();
+  const [tutor, setTutor] = useState<number | null>(null);
+
+  const closeRebookingDialog = useCallback(() => {
+    setTutor(null);
+  }, []);
+
+  const openRebookingDialog = useCallback((tutorId: number) => {
+    setTutor(tutorId);
+  }, []);
```
Comment (neuodev) [inline apps/nova/src/components/dashboard/PastLessonsWrapper.tsx:45]: I belive the code will be more clean if we removed these two functions `closeRebookingDialog` and `openRebookingDialog` and just `setTutor` directly. WDYT?

File: apps/nova/src/components/dashboard/PastLessonsWrapper.tsx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      />
+
+      {!lessonsQuery.query.isFetching ? (
+        <InView
+          as="div"
+          onChange={(inView) => {
+            if (inView && lessonsQuery.query.hasNextPage) lessonsQuery.more();
+          }}
+        />
+      ) : null}
```
Comment (neuodev) [inline apps/nova/src/components/dashboard/PastLessonsWrapper.tsx:87]: Nice!

File: apps/nova/src/components/dashboard/PastLessonsWrapper.tsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      <PastLessonsTable
+        tutorsRoute={Route.Tutors}
+        onRebook={openRebookingDialog}
+        lessons={lessons}
+        loading={lessonsQuery.query.isPending || lessonsQuery.query.isLoading}
+        error={lessonsQuery.query.isError}
+        retry={lessonsQuery.query.refetch}
+      />
+
+      {!lessonsQuery.query.isFetching ? (
```
Comment (neuodev) [inline apps/nova/src/components/dashboard/PastLessonsWrapper.tsx:80]: Also, remember not to show this `<InView` component if their is no more lessons.

File: apps/nova/src/components/dashboard/UpcomingLessonsSummaryWrapper.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    userOnly: true,
+    after: now,
+  });
+  const lessons = useMemo(
+    () => organizeUpcomingLessons(lessonsQuery.list),
+    [lessonsQuery.list]
+  );
+  return (
+    <UpcomingLessonsSummary
+      loading={lessonsQuery.query.isLoading || lessonsQuery.query.isPending}
```
Comment (neuodev) [inline apps/nova/src/components/dashboard/UpcomingLessonsSummaryWrapper.tsx:38]: Let's just show the spinner, incase of the `loading`. We don't want to distrub the layout of the page in case of re-fetching. The re-fetched data can be rendered without a loading indicator. 

```suggestion
      loading={lessonsQuery.query.isLoading}
```


This also applies to other page parts. 
- Chat summary. 
- Student overview.
- Past lessons table.

File: apps/nova/src/components/dashboard/ChatSummaryWrapper.tsx [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+export const ChatSummaryWrapper = () => {
```
Comment (neuodev) [inline apps/nova/src/components/dashboard/ChatSummaryWrapper.tsx:26]: I am not a fan of the word `...Wrapper`. I believe we can do better. We can just name it `ChatSummary`. For the original `ChatSummary` we can just call it the `Summary`. 

```tsx
import {ChatSummary as Summary} from '@litespace/luna/Chat'
export const ChatSummary = () => {
    return <Sumary />
}
```

This also applies to the file names and other components located in `components/dashboard/*`.

File: apps/nova/src/components/dashboard/ChatSummaryWrapper.tsx [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  const { user } = useUser();
+  const rooms = useFindUserRooms(user?.id, { size: 4 });
+  const organizedRooms = useMemo(() => organizeRooms(rooms.list), [rooms.list]);
+
+  return (
+    <ChatSummary
+      loading={rooms.query.isLoading || rooms.query.isPending}
```
Comment (neuodev) [inline apps/nova/src/components/dashboard/ChatSummaryWrapper.tsx:33]: They are the same I think. 
```suggestion
      loading={rooms.query.isLoading}
```

File: apps/dashboard/src/components/Students/StatsContent.tsx [hunk 12]
```diff
@@ -64,15 +64,21 @@ const StatsContent: React.FC<{
...
+        value: Duration.from(data.minutes.ratified.toString()).format({
+          language: "ar",
+        }),
       },
       {
         label: intl("stats.student.time.canceled"),
-        value: Duration.from(data.minutes.canceled.toString()).format("ar"),
+        value: Duration.from(data.minutes.canceled.toString()).format({
+          language: "ar",
+        }),
```
Comment (neuodev) [inline apps/dashboard/src/components/Students/StatsContent.tsx:81]: This is now how it should be done. Don't update the `Duration` class. Just write a local function to achive this here. Don't forget to revert the changes done in the `Duration` class.

**Review by neuodev — CHANGES_REQUESTED** [review]

File: apps/nova/src/components/dashboard/PastLessonsWrapper.tsx (hunk 6 above)
Comment (mostafakamar2308) [reply apps/nova/src/components/dashboard/PastLessonsWrapper.tsx:45]: yea, it's cleaner. but it will be more performant if we memoized the functions, WDYT?

File: apps/nova/src/components/dashboard/ChatSummaryWrapper.tsx (hunk 10 above)
Comment (mostafakamar2308) [reply apps/nova/src/components/dashboard/ChatSummaryWrapper.tsx:26]: Nice solution 🚀

File: apps/nova/src/components/dashboard/ChatSummaryWrapper.tsx (hunk 11 above)
Comment (mostafakamar2308) [reply apps/nova/src/components/dashboard/ChatSummaryWrapper.tsx:33]: Ok So, I've researched a bit more about this:
- `isPending`: there's no data and a query hasn't been made yet.
- `isFetching`: the query is being fetched
- `isLoading`: there's no data, and there's a query currently being fetched. Notice that it's a combination of the above.

it's recommended to use `isPending` [here](https://github.com/TanStack/query/discussions/6297#discussioncomment-7467010)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-a4juychyv-litespace-core.vercel.app) 
🚀 [Luna](https://luna-e9tjbocsk-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-nko8las9e-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-63jc8f1j1-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-ljcogqbx6-litespace-core.vercel.app) 
🚀 [Luna](https://luna-37ycy2wj3-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-6anltqda2-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-fzjlv2xwr-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-5zydu6abc-litespace-core.vercel.app) 
🚀 [Luna](https://luna-89jchmton-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-g6vwhd52z-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-lntnl9ti6-litespace-core.vercel.app)

File: apps/nova/src/components/dashboard/PastLessons.tsx [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        id: lesson.lesson.id,
+        start: lesson.lesson.start,
+        duration: lesson.lesson.duration,
+        tutor: {
+          id: teacher?.userId || 0,
+          name: teacher?.name || null,
+          imageUrl: teacher?.image || null,
+        },
+      };
+    }) || []
```
Comment (neuodev) [inline apps/nova/src/components/dashboard/PastLessons.tsx]:
```suggestion
    });
```

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-plhkw7837-litespace-core.vercel.app) 
🚀 [Luna](https://luna-6aeqafwri-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-grer4rjv9-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-cutwllrhz-litespace-core.vercel.app)
